config/config.example.yml

```diff
@@ -42,26 +44,67 @@ services:
+      Geoserver: # Service name, which should also be found in the services list at the top of this file
+        # Resource key (ex.: Geoserver_workspace): Custom name to represent a resource path.
+        # It should correspond to the names used in the `permission_mappings` section below.
+        # It contains the list of each segment of its resource path, with their corresponding name and type.
+        # The name of a segment can either contain a string name, a `SINGLE_TOKEN` `*` (to represent
+        # a single repetition of any name with the corresponding type) or a `MULTI_TOKEN` `**` (to
+        # represent any number (0 or more) of names that fit with the corresponding type). Note that tokens can
+        # only be found at the end of the path, so no named string should follow a token.
+        # Also, a MULTI_TOKEN can only be used one time in each list of path segments.
+        #
+        # The tokens are useful to know the types of any `suffix` path. A `suffix` path represents the segments
+        # at the end of a path that are not named in the config below. For example, with the config below, if we have
+        # an input resource path `/api/workspaces/private/dir1/dir2/file` that matches with the `Geoserver_workspace`
+        # resource key, we have a `suffix` path `/dir1/dir2/file`.
+        # Also, this key would need to sync permissions with the `Thredds_workspace` resource key, considering the
+        # `permission_mappings` defined below. The `Thredds_workspace` would use the path
+        # `/catalog/birdhouse/workspaces/private/` combined with the suffix `/dir1/dir2/file`.
+        # The types of each segment of the target resource path `/catalog/birdhouse/workspaces/private/dir1/dir2/file`
+        # would be deduced from the `Thredds_workspace` config below.
+        Geoserver_workspace:
+          - name: api
+            type: service
+          - name: workspaces
+            type: workspace
+          - name: private
+            type: workspace
+          - name: "**"
+            type: workspace
+      Thredds:
+        Thredds_workspace:
+          - name: catalog
+            type: service
+          - name: birdhouse
+            type: folder
+          - name: workspaces
+            type: folder
+          - name: private
+            type: folder
+          - name: "**"
+            type: folder
+          - name: "*"
+            type: file
     # [Required] This section defines an array of permissions mapping between services
-    # In this example, for every Geoserver's `read` permission, a `read` and `browse` permissions will be created in
-    # Thredds. If a `browse` permission is created in Thredds a `read` permission will be added to `Geoserver`.
+    # In this example, for every Geoserver_workspace's `read` permission, a `read` and `browse`
+    # permissions will be created in Thredds_workspace.
+    # If a `browse` permission is created in Thredds_workspace a `read` permission
```

Comment:
Yes, both `read` and `browse` in Thredds would generate a `read` in GeoServer. I will refine the comment.

For the permissions of a `ServiceGeoserver`, should it rather be one of those : `CreateStoreQuery`, `DescribeFeatureType`, `DescribeLayer`, `DescribeProcess`, etc.? It's the permissions I see if I create a geoserver service. Please confirm, and I'll change the config example to use one of those instead.

Reply:
This discussion brings another question.
For an example mapping : 
```
      - Geoserver_workspace:
          - read
        Thredds_workspace:
          - read
          - browse
```
If a webhook was received for a created `read` permission on `Thredds_workspace`, the initial cowbird implementation would only create a `read` permission for the `Geoserver_workspace`, and would ignore the other permissions found under `Thredds_workspace`. It would only update permissions found in other mapped resources.

@dbyrns Should it rather update every other possible permission, which would include the `browse` permission under `Thredds_workspace`? Let me know and I will update the PR.

Reply:
Indeed. I think in that case a `browse` should be equivalent to `GetMetadata` for `workspace`, and `DescribeFeatureType` for `file`. The `read` would be `GetMap`, `GetFeatureInfo`, `GetFeature`, `GetLegendGraphic`, `DescribeLayer`.

I'm not sure about the `CreateStoreQuery` and `Transaction` permissions (what they actually imply in GeoServer?). 
@f-PLT can maybe confirm which are metadata-related and which are actual data-getters.

It could make sense to grant `browse` (and corresponding GeoServer permissions) access for a resource that was granted `read`, since `read` is somewhat a superset of `browse`. On the other hand, there is a high possibility `browse` permissions are already granted access (via a parent recursive permission) to let users find available resources without immediate access until an agreement was completed. In this case, the over-creation of redundant `browse` on each resource will gradually slow down access validation (especially if requesting many layers simultaneously). I think it would be better to keep them separate, and that would simply mean Cowbird should receive 2 distinct sync operations (one for `browse` and one for `read`) if both were applied for a given resource. Mixing inter-permission logic with syncs might make it more convoluted to understand which access is actually granted when modifying them in Magpie.

Reply:
I'll look in to it, but right out of the gate, I think `Transaction` is about editing a vector feature type (creating, updating and deleting as per the documentation), so it's definitely a write operation.

The `CreateStoredQuery` is a POST request for saving a WFS query to the server, which can then be fetch and executed in a `GetFeature` request.

Reply:
Perfect. So both of these are `write` in Thredds context.

Reply:
> @dbyrns Should it rather update every other possible permission, which would include the browse permission under Thredds_workspace? Let me know and I will update the PR.

We only apply the synch for the permission that triggers the webhook. Any previous permission should already has been handled and if not, a full resynch should do that once implemented.

Reply:
If both `read` and `browse` should yied a `read` on the other service it's easy, but the other way around would cause to set both permissions... But like @fmigneault said, it could be redundant. 

- Maybe we could set a permission only if it's not already an effective permission?
- Maybe we could indicate which of multiple permissions should be set only. Meaning than a `read` set on geoserver could mean that we only set a `read` on thredds and not a `browse`?
- Maybe we could do both... Add a flag to a permission indicating that this one should be set only if not already effective for this user. Or a flag indicating that this one should be the only one to set from the permission list.

Reply:
Although I don't expect permissions to change often, technically adding the permission based on effective access could result in sudden lost access to that resource. If a permission is granted *explicitly* for a user, he should not depend on some inherited group to preserve that access. The examples are using the `user`, but we must remember that they apply to `group` as well. Effective access does not apply for groups. 

I think that most cases like `browse` involve some kind of higher-level container hierarchy (a directory, a workspace, etc.) and those are fairly constant. We don't need to apply these permissions individually per resource, and those should no receive permission sync too often.   I would tend to limit sync only for similar concepts (`read file <=> getfeature layer`). 

This is a pre requirement to have bidirectional mappings as the config currently enforces. If this is not viable and we start having too many flags to control what gets created, then another structure should be defined.

Reply:
I tried an example using the ** token. 
Here the code will eventually check that every 'formatting keys' (user and synched_file) exist in both mapped resource, and that both have a ** token. The subpath matching with the ** token in the first resource will be used to replace the ** token in the second resource.


```
services:
    Nginx: {}
    Thredds: {}
sync_permissions:
    user_workspace:
        services:
            Nginx:
                nginx_workspace:
                    - name: nginx
                      type: service
                    - name: workspaces
                      type: folder
                    - name: {user}
                      type: folder
                    - name: **
                      type: folder
                    - name: {synched_file}
                      type: file
            Thredds:
                thredds_workspace:
                    - name: catalog
                      type: service
                    - name: workspaces
                      type: folder
                    - name: {user}
                      type: folder
                    - name: **
                      type: folder
                    - name: subdir
                      type: folder
                    - name: {synched_file}
                      type: file
        permissions_mapping:
            # A read permission for nginx_workspace creates/deletes both read and browse permissions in 
            # thredds_workspace. And a read or a browse permission on thredds_workspace creates/deletes a read 
            # permission on nginx_workspace.
            - "nginx_workspace : read <-> thredds_workspace : [read,browse]"
            # A write permission on nginx_workspace creates/deletes a write permission on thredds_workspace, and vice 
            # versa.
            - "nginx_workspace : write <-> thredds_workspace : write"
```

Reply:
> if one has a '**' token, the other should have one too.

Only if bidirectional. Unidirectional ` ** -> ` could be allowed. Other way around ` -> **` is invalid, for same reasons and the named tokens.

Reply:
Sorry for beeing late in the game, but I see a small remaining issue with what @ChaamC said.

```
# [...] And a read or a browse permission on thredds_workspace creates/deletes a read 
# permission on nginx_workspace. - "nginx_workspace : read <-> thredds_workspace : [read,browse]"
```

I'm not sure that a `read` **OR** `browse` on `thredds_workspace` should delete a `read` permission on `nginx_workspace`.
It could be harder to implement, but the good way to do it would probably be to delete the `read` permission on `nginx_workspace` only if both `read` **AND** `browse` permissions have beed removed.

For multiples permissions, the rule would be :  

- If **ANY** permission of a set is **created**, the synch is performed on the other side and all of the permissions are created.
- If the **LAST** permission of a set is **deleted** (None of them exists anymore), the synch is performed on the other side and all of the permissions are deleted.
- Indeed, by "the synch is performed", I mean if the arrow allows it.

Finally, as we said, the config could be written in differents ways, but the internal object should be consistent and recongnise permissions set even when they are not explicitely stated. I said that because if I write two rules like this : 
```
- "nginx_workspace : read <-> thredds_workspace : read"
- "nginx_workspace : read <-> thredds_workspace : browse"
```
I would still expect the `nginx_workspace : read` permission to be deleted only after both `thredds_workspace` `read` and `browse` permissions have been deleted and not before.

Reply:
I'm not sure I agree with this interpretation of delete in nginx only if all were deleted in thredds. That means as long as I have a browse somewhere, I get elevated access to read files from nginx I shouldn't normally have access to. Also, browse could be applied on a parent resource (eg Directory), so it might never actually be deleted, meaning read on Files will always remain on nginx side.

The way browse is designed, it should be only to fetch metadata about the file, so read on both services should be blocked if it was deleted on any. Read applied on a thredds file should elevate browse as well (for consistency), but removing browse causes no action the other way around. 

Also, in this case, it just so happens that browse/read correspond to somewhat similar "concepts" of another service's read, but that kind of mapping does not apply for other permissions like getcapabilities/describeprocess/execute.

Reply:
I agree, a parent permission should not prevent any deletion, I'm really talking about a resource specific permissions.

The rule we want to implement is `If A OR B Then C`.

1. Create A, Then create C
2. Create B, Then create C # Good already done !
3. Delete A, Then delete C # !?! If C is deleted at this point the rule A OR B Then C is no more valid, we have B, but no C.
4. Delete B, Then delete C # At this point the rule is valid because A and B have been removed.

What it means with the thredds example is that if I add a `read` (A) or a `browse` (B) permissions directly on a thredds resource, a `read` permission (C) is added to nginx. We're not talking about a parent's `browse` permission. So, for that specific ressource, as long as a `read` (A) or `browse` (B) permission exists, it does not make sense to remove the nginx `read` (C) permission.

Reply:
Ok. I agree about the AND requirement prior to deleting C in that case.

Regarding the specific permissions. Does it make sense to grant nginx read when adding (only) thredds browse?
Imagine that thredds read was not applied for the user. This means we explicitly allow the user to be able see metadata about the file (eg: to let him know it exists, request access to it, etc.?), but he cannot download it yet (eg: pending agreement, T&C validation etc.). In that case, the user could just bypass thredds and retrieve the file anyway via nginx.

Reply:
Yes you're right, we were trying to find examples, but the final config should not include a browse permission. It does not make sense in any direction. We don't want to set browse permissions everywhere when a read is already given. And we cannot allow a read on nginx because of a browse only permission.

Reply:
I just found the time to read your latest messages. If I understand, I would have to implement the mentioned rule `If A OR B Then C`.
Also, should I change the example given to avoid any confusion, since it seems not to be a desired case used for the final config. Is there a better type of permission I could use to showcase the example of 1 permission synchronized with 2 other permissions from another service?

Reply:
Maybe something using the `ServiceGeoserver`? All permissions doing `GetFeature`, `GetMap`, etc. could be considered like "reading" the same data but with different OWS representations/services. Creating/deleting those permissions in sync when similar "reading" intention is applied for a user could make sense.